### PR TITLE
APEXCORE-493 checkpoint purging from Streaming container

### PR DESCRIPTION
**Pull request by sandeshh**
Checkpoint is stored by StreamingContainer and will be cleaned by the StreamingContainer.

### CODE REVIEW

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 1]
```diff
@@ -2123,7 +2122,6 @@ public void updateRecoveryCheckpoints(PTOperator operator, UpdateCheckpointsCont
     for (PTOperator groupOper : groupOpers) {
       // checkpoint frozen during deployment
       if (!pendingDeploy || ctx.recovery) {
-        // remove previous checkpoints
         Checkpoint c1 = Checkpoint.INITIAL_CHECKPOINT;
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:2125]: The INITIAL_CHECKPOINT seems to be included into the purge list. If it is excluded later, how this exclusion happens?

**Review by vrozov — APPROVED** [review]
Overall looks OK to me (see comments regarding INITIAL_CHECKPOINT). @tweise Please take a look.

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java [hunk 2]
```diff
@@ -776,6 +776,26 @@ public void processHeartbeatResponse(ContainerHeartbeatResponse rsp)
         bufferServer.purge(lastCommittedWindowId - 1);
       }
 
+      for (Node<?> node : nodes.values()) {
+
+        if (!node.context.stateless) {
+          StorageAgent storageAgent = node.context.getValue(OperatorContext.STORAGE_AGENT);
+          try {
+            for (long windowId : storageAgent.getWindowIds(node.getId())) {
```
Comment (vrozov) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: sort by windowId and break out of the loop once windowId >= lastCommittedWindowId

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 1 above)
Comment (sandeshh) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java:2125]: It is overwritten in the line 2131, before purge.

File: engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java (hunk 2 above)
Comment (sandeshh) [inline engine/src/main/java/com/datatorrent/stram/engine/StreamingContainer.java]: There is no advantage to sorting here, as traversal is happening only once.

**Comment by vrozov** [comment]
After looking into details of how checkpoints are purged I think that there will be a minimal benefit, if any, in delegating delete of checkpoints to each operator's container. As an application master does not send requests to delete checkpoints to each container and directly calls Store Agent that actually performs the delete, benefits of delegating delete of checkpoints to operator's containers are questionable. Additionally, when a DAG has disconnected parts, it looks that the new implementation would delete less checkpoint compared to the old implementation.

**Comment by vrozov** [comment]
@tweise @PramodSSImmaneni Do you see any benefit of delegating checkpoints purge to containers from AM?

**Comment by pramodin** [comment]
Not at the moment. In future, if we see storage implementations where this is a heavy operation we can think about it. I also worry it introduces an element of uncertainty when the purge happens, for example, if the operator had a failure at that same point of time, the purge will have to wait for the next committed cycle.

**Comment by sandeshh** [comment]
Thanks for the feedback. I will close the PR for now.

**Pull request closed without merging**